### DISCUSSION TITLE
What is the router like?

### DISCUSSION THREAD

**Discussion by ghost** [beeware/briefcase#2047] [Q&A]
I want to put a config file (`config.json`), but I can't read it with 
```python
with open("config.json', 'r') as f:
```
-->`ERROR: file not found`
The file in under the same dictionary of `app.py`. 

what is the router like and how to read and write files?

**Comment by freakboy3742** [comment]
I'm not sure what you mean by "the router" - Briefcase doesn't have any concept of a router.

However, the problem you're hitting is well understood in Briefcase-packaged apps - it's an issue with the current working directory.

When you `open("config.json")`, you're not explicitly giving the location of a file - you're *implicitly* asking to open a file named `config.json` *in your current working directory*. When you're running Python from a command line, that location is obvious - it's the location where you started the interpreter. But in a packaged app, it's not as obvious - what is your current working directory when you double click on an icon? Briefcase will *deliberately* set the working directory to the user's home directory *specifically* to highlight bugs like this one.

The fix is to *not* be implicit - and provide a fully qualified path to the file. Don't `open("config.json")` - instead, `open("/full/path/to/config.json")`.

Determining `/full/path/to` is the next problem - and the solution depends on what toolkit you're using.

`__file__` can be used in any Python code to report the full file that the code is currently in. You can then use Pathlib to get the directory that contains the current python file with `Path(__file__).parent`; and then construct the path to the config file from that root.

If you're using Toga, a Toga app provides [`App.paths`](https://toga.readthedocs.io/en/stable/reference/api/app.html#toga.App.paths) that gives you access to a range of useful filesystem roots (in Pathlib format) that you can use to construct a full path.

**Reply by ghost** [reply]
thanks, it work. 
my router means how the programme look for the file.
